### PR TITLE
tp: Improved handling of low-queue state

### PR DESCRIPTION
**Pull request by robEllenberg**
The problem: Optimization falters when the queue only has a few segments
in it. Specifically, we can't be sure of the total segment length until
there is another segment after it (due to blending). The approach in the
past was very conservative, and basically aborted optimization entirely
if the queue had less than 3 segments in it.

This commit addresses this problem somewhat with 2 major changes:
1) Allow blending on segments already in progress, provided that a few
conditions are true (not doing parabolic blending already, less than
1/2 of the segment is complete).
2) Since a segment will never have more than 1/2 of it's length consumed
in a blend, the worst-case final velocity of the previous segment is the
maximum velocity that we can reach and still stop at the halfway point.

With these fixes, we don't have to fall back to parabolic blending at
the beginning and end of programs (at least not as frequently), so
execution is much more consistent. The trade-off with this approach is
that the last segment before a queue-buster or the end of a program has
slightly poorer blending performance.

### CODE REVIEW

File: src/emc/tp/tp.c [hunk 1]
```diff
@@ -697,24 +697,29 @@ STATIC double tpCalculateTriangleVel(TP_STRUCT const * const tp, TC_STRUCT * con
...
+    return fmin(triangle_vel, max_vel);
 }
 
+
 /**
- * Initialize a blend arc from its parent segments.
- * This copies and initializes properties from the previous and next segments to
+ * Initialize a blend arc from its parent lines.
+ * This copies and initializes properties from the previous and next lines to
+>>>>>>> 45bec86... tp: Improved handling of low-queue state
```
Comment (zultron) [inline src/emc/tp/tp.c]: This merge marker is unintentional?

**Comment by mhaberler** [comment]
boy, things are moving quick these days ;)

**Comment by strahlex** [comment]
Can it be that this solves my problems with the motion pins? The first few prints worked without any problems.

**Comment by robEllenberg** [comment]
Hi Alex,

That would be a nice side effect, certainly! Are you referring to the 4th
axis issues we were having before (the RT hangups), or was this a different
issue?

-Rob

On Mon, Mar 2, 2015 at 10:25 AM, Alexander Rössler <notifications@github.com

**Comment by strahlex** [comment]
Hi Rob,

No, I was referring to the problem I had with the synchronized analog
outputs. But, I will find out if it solves the problems with the fourth
axis too.

Robert Ellenberg writes:

> Hi Alex,
>
> That would be a nice side effect, certainly! Are you referring to the 4th
> axis issues we were having before (the RT hangups), or was this a different
> issue?
>
> -Rob
>
> On Mon, Mar 2, 2015 at 10:25 AM, Alexander Rössler <notifications@github.com
>> wrote:
>
>> Can it be that this solves my problems with the motion pins? The first few
>> prints worked without any problems.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/machinekit/machinekit/pull/505#issuecomment-76731139>.
>>
>
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/machinekit/machinekit/pull/505#issuecomment-76804978

-- 
Alexander Rössler

**Pull request merged**